### PR TITLE
Apply Kotlin validation in INPUT requests DTOs 

### PR DESCRIPTION
**Pull request by rkonoplev**
[EPMLABSBRN-891](https://github.com/Brain-up/brn/issues/2052)

**Description**:

use Kotlin validation in input request DTOs, refactor Java validation intro Kotlin by example

**Details**:

like in UserAccountCreateRequest: @field:NotEmpty

instead of simple @notblank as in Java(as it is not work in Kotlin)

### CODE REVIEW

File: src/main/kotlin/com/epam/brn/dto/HeadphonesDto.kt [hunk 1]
```diff
@@ -6,8 +6,8 @@ import javax.validation.constraints.NotBlank
 
 data class HeadphonesDto(
     var id: Long? = null,
-    @NotBlank
-    var name: String = "",
+    @field:NotBlank
+    var name: String,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/HeadphonesDto.kt:10]: Are you sure you need to remove the default value?

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt [hunk 2]
```diff
@@ -2,14 +2,14 @@ package com.epam.brn.dto
 
 import com.epam.brn.model.ExerciseType
 import javax.validation.constraints.NotBlank
+import javax.validation.constraints.NotNull
 
 data class SeriesDto(
-    @NotBlank
-    val group: Long? = null,
+    @field:NotNull
+    val group: Long,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:9]: I think it's better to leave the nullable for now. This is beyond the scope of the task

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt [hunk 3]
```diff
@@ -2,14 +2,14 @@ package com.epam.brn.dto
...
     val id: Long?,
-    @NotBlank
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:10]: Why did I delete this?

File: src/main/kotlin/com/epam/brn/dto/SignalTaskDto.kt [hunk 4]
```diff
@@ -4,7 +4,7 @@ import com.fasterxml.jackson.annotation.JsonIgnore
 
 data class SignalTaskDto(
     val id: Long? = null,
-    @JsonIgnore
+    @field:JsonIgnore
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/SignalTaskDto.kt:7]: there is no need to change here. This is not validation

File: src/main/kotlin/com/epam/brn/dto/WordsGroupSeriesTaskDto.kt [hunk 5]
```diff
@@ -7,7 +7,7 @@ import com.fasterxml.jackson.annotation.JsonIgnore
 data class WordsGroupSeriesTaskDto(
     val id: Long,
     val exerciseType: ExerciseType,
-    @JsonIgnore
+    @field:JsonIgnore
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/WordsGroupSeriesTaskDto.kt:10]: there is no need to change here. This is not validation

File: src/main/kotlin/com/epam/brn/dto/WordsSeriesTaskDto.kt [hunk 6]
```diff
@@ -6,7 +6,7 @@ import com.fasterxml.jackson.annotation.JsonIgnore
 data class WordsSeriesTaskDto(
     val id: Long,
     val exerciseType: ExerciseType,
-    @JsonIgnore
+    @field:JsonIgnore
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/WordsSeriesTaskDto.kt:9]: there is no need to change here. This is not validation

File: src/main/kotlin/com/epam/brn/dto/request/AudiometryHistoryRequest.kt [hunk 7]
```diff
@@ -9,18 +9,18 @@ import javax.annotation.Nullable
...
     var startTime: LocalDateTime,
     var endTime: LocalDateTime?,
-    @NotNull
+    @field:NotNull
     var executionSeconds: Int?,
-    @NotNull
+    @field:NotNull
     var tasksCount: Short,
-    @Nullable
+    @field: Nullable
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/AudiometryHistoryRequest.kt:21]: And here it should definitely be @Nullable

File: src/main/kotlin/com/epam/brn/dto/request/ExerciseRequest.kt [hunk 8]
```diff
@@ -6,6 +6,7 @@ import com.fasterxml.jackson.annotation.JsonProperty
 
 @JsonInclude
 data class ExerciseRequest @JsonCreator constructor(
-    @JsonProperty("ids")
+    @param:JsonProperty("ids")
+    @get:JsonProperty("ids")
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/ExerciseRequest.kt:10]: there is no need to change here. This is not validation

File: src/main/kotlin/com/epam/brn/dto/request/LoginDto.kt [hunk 9]
```diff
@@ -9,6 +9,7 @@ data class LoginDto(
     val grant_type: String = "password",
     @field:NotBlank(message = "{validation.field.email.blank}")
     @field:Email(message = "{validation.field.email.invalid-format}")
+    val email: String,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/LoginDto.kt:12]: why did you write this field here?

File: src/main/kotlin/com/epam/brn/model/Series.kt [hunk 10]
```diff
@@ -40,7 +40,7 @@ data class Series(
     )
 
     fun toDto() = SeriesDto(
-        group = exerciseGroup.id,
+        group = exerciseGroup.id!!,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/model/Series.kt:43]: see the comment above in SeriesDto. kt

File: src/main/kotlin/com/epam/brn/dto/HeadphonesDto.kt (hunk 1 above)
Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/dto/HeadphonesDto.kt:10]: This case was approver earlier.

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt (hunk 2 above)
Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:9]: This case was approver earlier.

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt (hunk 3 above)
Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:10]: This case was approver earlier.

File: src/main/kotlin/com/epam/brn/dto/request/LoginDto.kt (hunk 9 above)
Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/dto/request/LoginDto.kt]: This case was approver earlier.

File: src/main/kotlin/com/epam/brn/model/Series.kt (hunk 10 above)
Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/model/Series.kt:43]: This case was approved earlier.

File: src/main/kotlin/com/epam/brn/controller/AdminController.kt [hunk 11]
```diff
@@ -183,4 +189,18 @@ class AdminController(
     ): ResponseEntity<BaseSingleObjectResponseDto> =
         ResponseEntity.status(HttpStatus.CREATED)
             .body(BaseSingleObjectResponseDto(data = subGroupService.addSubGroupToSeries(subGroupRequest, seriesId)))
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException::class)
+    fun handleValidationExceptions(ex: MethodArgumentNotValidException): Map<String, String?>? {
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/controller/AdminController.kt:195]: why is it here? no need to add your debug code to the PR

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt (hunk 2 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:9]: return the nullable. We decided to make variables with an annotation @NotNull nullable

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 12]
```diff
@@ -2,16 +2,21 @@ package com.epam.brn.dto.request
...
-import javax.validation.constraints.NotBlank
-import javax.validation.constraints.NotNull
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
 
-data class SubGroupRequest(
-    @NotBlank
+data class SubGroupRequest @JsonCreator internal constructor (
+    @get:JsonProperty(value = "first-property", required = true)
+    @param:JsonProperty(value = "first-property", required = true)
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:10]: these annotations are not needed here. return @field:NotBlank

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 13]
```diff
@@ -2,16 +2,21 @@ package com.epam.brn.dto.request
...
     var name: String,
-    @NotNull
+    @get:JsonProperty(value = "second-property", required = true)
+    @param:JsonProperty(value = "second-property", required = true)
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:13]: same

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 14]
```diff
@@ -2,16 +2,21 @@ package com.epam.brn.dto.request
...
     var level: Int,
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:14]: Make field nullable

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 15]
```diff
@@ -2,16 +2,21 @@ package com.epam.brn.dto.request
...
-    @NotBlank
+    @get:JsonProperty("third-property", required = true)
+    @param:JsonProperty("third-property", required = true)
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:16]: same

File: src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt [hunk 16]
```diff
@@ -2,16 +2,21 @@ package com.epam.brn.dto.request
...
     var code: String,
+    @get:JsonProperty("fourth-property")
+    @param:JsonProperty("fourth-property")
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/SubGroupRequest.kt:19]: same

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 17]
```diff
@@ -511,6 +515,79 @@ class AdminControllerIT : BaseIT() {
             .andExpect(jsonPath("$.data.name").value(subGroupRequest.name))
     }
 
+    @Test
+    @Throws(Exception::class)
+    fun `when post request to subGroup and invalid name in subGroup then correct response`() {
```
Comment (Dmitriy1Fokin) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:520]: these three tests can be combined into one. One endpoint one test

File: src/main/kotlin/com/epam/brn/dto/request/UserAccountCreateRequest.kt [hunk 18]
```diff
@@ -15,7 +14,7 @@ const val VALID_EMAIL_ADDRESS_REGEX_WITH_EMPTY_SPACES_ACCEPTANCE: String =
 
 @JsonInclude(JsonInclude.Include.NON_NULL)
 data class UserAccountCreateRequest(
-    @field:NotEmpty(message = "{validation.field.fullName.empty}")
+    @field:NotBlank(message = "{validation.field.fullName.empty}")
```
Comment (Dmitriy1Fokin) [inline src/main/kotlin/com/epam/brn/dto/request/UserAccountCreateRequest.kt:17]: change message

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 19]
```diff
@@ -511,6 +515,79 @@ class AdminControllerIT : BaseIT() {
...
+        // GIVEN
+        val subGroupRequest =
+            """{"name":"","code":"12345","level":"1","description":"Test description" }"""
+        val existedSeries = insertSeries()
```
Comment (Dmitriy1Fokin) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:524]: any id is needed for the test. you don't need to create existedSeries

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 20]
```diff
@@ -511,6 +515,79 @@ class AdminControllerIT : BaseIT() {
...
+            MockMvcRequestBuilders
+                .post("$baseUrl/subgroup")
+                .param("seriesId", seriesId.toString())
+                .content(requestBody)
+                .contentType(MediaType.APPLICATION_JSON)
+        )
+        // THEN
+        response
+            .andExpect(MockMvcResultMatchers.status().isBadRequest)
+            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
```
Comment (Dmitriy1Fokin) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:539]: you need to check the response body. It should contain the names of fields with errors

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 21]
```diff
@@ -511,6 +515,79 @@ class AdminControllerIT : BaseIT() {
...
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun `when post request to subGroup and invalid code in subGroup then correct response`() {
```
Comment (Dmitriy1Fokin) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:544]: the same comments as for the test above

File: src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt [hunk 22]
```diff
@@ -511,6 +515,79 @@ class AdminControllerIT : BaseIT() {
...
+                .contentType(MediaType.APPLICATION_JSON)
+        )
+        // THEN
+        response
+            .andExpect(MockMvcResultMatchers.status().isBadRequest)
+            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+    }
+
+    @Test
+    fun `when post request to subGroup and invalid level in subGroup then correct response`() {
```
Comment (Dmitriy1Fokin) [inline src/test/kotlin/com/epam/brn/integration/AdminControllerIT.kt:567]: the same comments as for the test above

**Review by Dmitriy1Fokin — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/epam/brn/dto/SeriesDto.kt (hunk 2 above)
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/SeriesDto.kt:9]: as we discussed it should be Long? if it is used as request dto,
if only as a response - we can delete it

File: src/main/kotlin/com/epam/brn/dto/StudyHistoryDto.kt [hunk 23]
```diff
@@ -8,18 +8,18 @@ import javax.validation.constraints.NotNull
...
+    @field:NotNull
     var executionSeconds: Int,
-    @NotNull
+    @field:NotNull
     var tasksCount: Short,
-    @NotNull
+    @field:NotNull
     var replaysCount: Int, // -- count of replays.  --
-    @NotNull
+    @field:NotNull
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/StudyHistoryDto.kt:22]: same

File: src/main/kotlin/com/epam/brn/dto/request/AudiometryHistoryRequest.kt [hunk 24]
```diff
@@ -9,19 +9,19 @@ import javax.annotation.Nullable
...
-    @NotNull
-    var tasksCount: Short,
-    @Nullable
-    var rightAnswers: Int = 0,
-    @NotNull
-    var headphones: Long? = null,
+    @field:NotNull
+    var tasksCount: Short?,
+    @field:Nullable
+    var rightAnswers: Int,
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/request/AudiometryHistoryRequest.kt:22]: @field:Nullable
var rightAnswers: Int,

->

   var rightAnswers: Int?=0,

Comment (rkonoplev) [reply src/main/kotlin/com/epam/brn/dto/request/AudiometryHistoryRequest.kt]: Resolved

**Review by ElenaSpb — APPROVED** [review]

**Pull request merged**